### PR TITLE
[MAINTENANCE] Refactor `Table.head()` for `sqlalchemy`

### PR DESCRIPTION
**Pull request by Shinnnyshinshin**
# What does this PR do? 
* Clean up of `Table.head()` metric SqlAlchemy case
* Helper method for `pandas < 1.4`
* Unit-tests for `table.head()` metric. 
- [x] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [x] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [x] Code is linted - run `invoke lint` (uses `black` + `ruff`)
- [x] Appropriate tests and docs have been updated

### CODE REVIEW

File: great_expectations/expectations/metrics/table_metrics/table_head.py [hunk 1]
```diff
@@ -70,132 +74,75 @@ def _sqlalchemy(  # noqa: C901, PLR0912, PLR0913, PLR0915
...
-                )
+        if is_version_less_than(pd.__version__, "1.4.0"):
+            df = TableHead._sqlalchemy_head_pandas_less_than14(
+                selectable=selectable,
+                table_name=table_name,
+                execution_engine=execution_engine,
+                metric_value_kwargs=metric_value_kwargs,
+                n_rows=n_rows,
+            )
+            return df
```
Comment (Shinnnyshinshin) [inline great_expectations/expectations/metrics/table_metrics/table_head.py]: What was previously the `df = None` case handled by its own helper method.

File: great_expectations/expectations/metrics/table_metrics/table_head.py [hunk 2]
```diff
@@ -248,3 +195,84 @@ def _spark(  # noqa: PLR0913
...
+            with execution_engine.get_connection() as con:
+                df = pandas_read_sql(
+                    sql=selectable,
+                    con=con,
+                    execution_engine=execution_engine,
+                )
+        return df
+
+    @staticmethod
+    def _sqlalchemy_head_pandas_less_than14(
```
Comment (Shinnnyshinshin) [inline great_expectations/expectations/metrics/table_metrics/table_head.py:220]: The code here is mostly the same, just moved to its own helper method

File: great_expectations/self_check/util.py [hunk 3]
```diff
@@ -1224,9 +1224,8 @@ def build_sa_execution_engine(  # noqa: PLR0913
     if_exists: str = "replace",
     index: bool = False,
     dtype: Optional[dict] = None,
+    table_name: str = "test",
 ) -> SqlAlchemyExecutionEngine:
-    table_name: str = "test"
-
```
Comment (Shinnnyshinshin) [inline great_expectations/self_check/util.py:1229]: `table_name` can now be over-written if needed

File: great_expectations/expectations/metrics/table_metrics/table_head.py (hunk 2 above)
Comment (Shinnnyshinshin) [reply great_expectations/expectations/metrics/table_metrics/table_head.py:220]: I hope to retire this method if we bump our minimum pandas to greater than 1.4

File: great_expectations/expectations/metrics/table_metrics/table_head.py [hunk 4]
```diff
@@ -248,3 +190,83 @@ def _spark(  # noqa: PLR0913
...
+        """
+        stmt = sa.select("*").select_from(selectable)
+        fetch_all = metric_value_kwargs["fetch_all"]
+        if fetch_all:
+            sql = stmt.compile(
+                dialect=execution_engine.engine.dialect,
+                compile_kwargs={"literal_binds": True},
+            )
+        elif execution_engine.engine.dialect.name.lower() == GXSqlDialect.MSSQL:
+            # limit doesn't compile properly for mssql
```
Comment (anthonyburdi) [inline great_expectations/expectations/metrics/table_metrics/table_head.py:240]: Interesting, so for later pandas versions the limit compiles properly. Good find!

File: tests/expectations/metrics/table_metrics/test_table_head.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+@pytest.fixture
+def sqlalchemy_test_engine(sa):
+    return build_sa_execution_engine(
+        pd.DataFrame({"a": [1, 2, 3, 4], "b": [2, 3, 4, 5]}),
+        sa,
+        table_name="test_table",
+    )
+
+
+def test_head_table_name_not_fetch_all_2_rows(sqlalchemy_test_engine):
```
Comment (anthonyburdi) [inline tests/expectations/metrics/table_metrics/test_table_head.py]: Please add some decorators to these tests.

File: tests/expectations/metrics/table_metrics/test_table_head.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    assert res.shape == (n_rows, 2)
+    assert res.columns.tolist() == ["a", "b"]
+    assert res.values.tolist() == [[1, 2], [2, 3]]
+
+
+def test_head_no_table_name_not_fetch_all_2_rows(sqlalchemy_test_engine):
+    execution_engine: SqlAlchemyExecutionEngine = sqlalchemy_test_engine
+    table_head = TableHead()
+
+    table_name = None
```
Comment (anthonyburdi) [inline tests/expectations/metrics/table_metrics/test_table_head.py]: Sorry if I'm missing this - how is table name used in this test?

File: tests/expectations/metrics/table_metrics/test_table_head.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        metrics={},
+        runtime_configuration={},
+    )
+
+    assert res.shape == (n_rows, 2)
+    assert res.columns.tolist() == ["a", "b"]
+    assert res.values.tolist() == [[1, 2], [2, 3]]
+
+
+def test_head_table_name_is_anonymous_not_fetch_all(sqlalchemy_test_engine):
```
Comment (anthonyburdi) [inline tests/expectations/metrics/table_metrics/test_table_head.py]: Up to you, but most of these tests only change a parameter or two - could be a good candidate for a parametrized test. Param IDs can be used to describe what is being tested (e.g. in place of the test function name). There are a few examples in the codebase, e.g. `test_generate_table_name_with_expectation` or just search for the decorator `@pytest.mark.parametrize`

**Review by anthonyburdi — CHANGES_REQUESTED** [review]
Nice work! Just a few items to address.

File: tests/expectations/metrics/table_metrics/test_table_head.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+            "sqlite_batch_with_selectable_without_temp_table",
+            None,
+            True,
+            (4, 2),
+            ["a", "b"],
+            [[1, 2], [2, 3], [3, 4], [4, 5]],
+        ),
+    ],
+)
+def test_table_head_sqlite(
```
Comment (Shinnnyshinshin) [inline tests/expectations/metrics/table_metrics/test_table_head.py:139]: @anthonyburdi the tests have been parametrized :)

**Review by anthonyburdi — APPROVED** [review]
LGTM! Thanks for the synchronous review :)

**Pull request merged**